Client/Frontend/Shields/ShareShieldsController.swift

```diff
@@ -0,0 +1,278 @@
+    }
+
+    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
+        super.traitCollectionDidChange(previousTraitCollection)
+
+        if #available(iOS 13.0, *) {
+            if traitCollection.userInterfaceStyle != previousTraitCollection?.userInterfaceStyle {
+                applyTheme(theme)
+            }
+        }
+    }
+
+    // MARK: Internal
+    
+    private func doLayoutAndAddGestureRecognizers() {
+        let shareTypeViewEmail = ShareTypeView(type: .email)
+        let shareEmailGesture = UITapGestureRecognizer(target: self, action: #selector(shareClicked(_:)))
+        shareTypeViewEmail.addGestureRecognizer(shareEmailGesture)
+
+        let shareTypeViewTwitter = ShareTypeView(type: .twitter)
+        let shareTwitterGesture = UITapGestureRecognizer(target: self, action: #selector(shareClicked(_:)))
+        shareTypeViewTwitter.addGestureRecognizer(shareTwitterGesture)
+
+        let shareTypeViewFacebook = ShareTypeView(type: .facebook)
+        let shareFacebookGesture = UITapGestureRecognizer(target: self, action: #selector(shareClicked(_:)))
+        shareTypeViewFacebook.addGestureRecognizer(shareFacebookGesture)
+
+        let shareTypeViewDefault = ShareTypeView(type: .default)
+        let shareDefaultGesture = UITapGestureRecognizer(target: self, action: #selector(shareClicked(_:)))
+        shareTypeViewDefault.addGestureRecognizer(shareDefaultGesture)
+
+        view.addSubview(contentStackView)
+
+        contentStackView.snp.makeConstraints {
+            $0.top.leading.trailing.equalTo(view.safeAreaLayoutGuide)
+        }
+
+        contentStackView.addStackViewItems(
+            .view(shareTypeViewEmail),
+            .view(shareTypeViewTwitter),
+            .view(shareTypeViewFacebook),
+            .view(shareTypeViewDefault)
+        )
+    }
+
+    // MARK: Themeable
+    
+    func applyTheme(_ theme: Theme) {
+        view.backgroundColor = theme.isDark ? BraveUX.popoverDarkBackground : UIColor.white
+
+        contentStackView.arrangedSubviews.forEach { subview in
+            if let typeView = subview as? ShareTypeView {
+                typeView.applyTheme(theme)
+            }
+        }
+    }
+
+    // MARK: Actions
+    
+    @objc fileprivate func shareClicked(_ sender: UITapGestureRecognizer) {
+        if let typeView = sender.view as? ShareTypeView {
+            switch typeView.shareType {
+                case .email:
+                    actionHandler?(.shareEmailClicked)
+                case .twitter:
+                    actionHandler?(.shareTwitterClicked)
+                case .facebook:
+                    actionHandler?(.shareFacebookClicked)
+                case .default:
+                    actionHandler?(.shareDefaultClicked)
+            }
+        }
+    }
+}
+
+// MARK: - ShareTypeView
+
+private class ShareTypeView: UIView, Themeable {
+
+    // MARK: ShareType
+    
+    enum ShareType {
```

Comment:
Thoughts on merging `ShareShieldsViewController.Action` and `ShareType` together? both have the same values,

and these `title` `icon` getters could be moved out of the enum and placed in this view

Reply:
This makes it easier and cleaner to track down, That View has it own local property working over the enum and this gives us flexibility so we can use it somewhere else.

Reply:
I understand the idea behind it but I prefer to keep them separate, that view is its own identity and it has types and properties according to these types. Other one represents an action event for a different place (a alternative to a protocol)